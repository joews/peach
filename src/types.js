// Type definitions used by the type checker
class TypeOperator {
  constructor (name, typeArgs = []) {
    this.name = name
    this.typeArgs = typeArgs
  }

  // polymorphic factory
  static of (name, typeArgs) {
    return new TypeOperator(name, typeArgs)
  }

  toString () {
    return this.name
  }
}

const NumberType = new TypeOperator('Number')
const StringType = new TypeOperator('String')
const BooleanType = new TypeOperator('Boolean')

class FunctionType extends TypeOperator {
  constructor (argType, returnType) {
    super('->', [argType, returnType])
  }

  static of (name, types) {
    const [argType, returnType] = types
    return new FunctionType(argType, returnType)
  }

  // getArgTypes () {
  //   return this.typeArgs.slice(0, -1)
  // }

  getArgType () {
    return this.typeArgs[0]
  }

  getReturnType () {
    return this.typeArgs[1]
  }

  toString () {
    const argType = this.getArgType()
    const returnType = this.getReturnType()

    let argString
    if (argType == null) {
      argString = '()'
    } else if (argType instanceof FunctionType || argType.instance instanceof FunctionType) {
      argString = `(${argType})`
    } else {
      argString = argType
    }

    return `${argString} -> ${returnType}`
  }
}

class ArrayType extends TypeOperator {
  constructor (argType) {
    super('Array', [argType])
  }

  static of (name, types) {
    return new ArrayType(types[0])
  }

  getType () {
    return this.typeArgs[0]
  }

  toString () {
    return `Array<${this.getType()}>`
  }
}

class TypeVariable {
  constructor () {
    this.id = TypeVariable.nextId ++
    this.name = null
  }

  toString () {
    if (this.instance) {
      return this.instance.toString()
    } else {
      if (!this.name) {
        this.name = String.fromCharCode(TypeVariable.nextName ++)
      }

      return this.name
    }
  }
}

// Factory for curried function types that take any number of arguments
function makeFunctionType (argTypes, returnType) {
  const [firstArgType, ...tailArgTypes] = argTypes

  if (tailArgTypes.length === 0) {
    return new FunctionType(firstArgType, returnType)
  } else {
    return new FunctionType(firstArgType, makeFunctionType(tailArgTypes, returnType))
  }
}

// TODO
// > Contextual names for TypeVariables, so that many tests in the same
//   JS process can start type variables at A
// > Make nice names for type variables once we have created a lot. E.g. after Z,
//   go to AA.
TypeVariable.nextId = 0
TypeVariable.nextName = 65

module.exports = {
  TypeVariable,
  TypeOperator,
  FunctionType,
  ArrayType,
  NumberType,
  StringType,
  BooleanType,
  makeFunctionType
}
