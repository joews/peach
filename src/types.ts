// Type definitions used by the type checker
export class Type {
  name: string

  constructor (name: string) {
    this.name = name
  }
}

export class TypeOperator extends Type {
  typeArgs: Array<Type>

  constructor (name: string, typeArgs: Type[] = []) {
    super(name)
    this.typeArgs = typeArgs
  }

  // polymorphic factory
  static of (name: string, typeArgs: Type[]) {
    return new TypeOperator(name, typeArgs)
  }

  toString () {
    return this.name
  }
}

export const NumberType = new TypeOperator('Number')
export const StringType = new TypeOperator('String')
export const BooleanType = new TypeOperator('Boolean')

export class FunctionType extends TypeOperator {
  constructor (argType: Type, returnType: Type) {
    super('->', [argType, returnType])
  }

  static of (name: string, types: Type[]) {
    const [argType, returnType] = types
    return new FunctionType(argType, returnType)
  }

  // getArgTypes () {
  //   return this.typeArgs.slice(0, -1)
  // }

  getArgType () {
    return this.typeArgs[0]
  }

  getReturnType () {
    return this.typeArgs[1]
  }

  toString () {
    const argType = this.getArgType()
    const returnType = this.getReturnType()

    let argString
    if (argType == null) {
      argString = '()'
    } else if (argType instanceof FunctionType || argType instanceof TypeVariable && argType.instance instanceof FunctionType) {
      argString = `(${argType})`
    } else {
      argString = argType
    }

    return `${argString} -> ${returnType}`
  }
}

export class ArrayType extends TypeOperator {
  constructor (argType: Type) {
    super('Array', [argType])
  }

  static of (name: string, types: Type[]) {
    return new ArrayType(types[0])
  }

  getType () {
    return this.typeArgs[0]
  }

  toString () {
    return `Array<${this.getType()}>`
  }
}

export class TupleType extends TypeOperator {
  constructor (types: Type[]) {
    super('Tuple', types)
  }

  static of (name: string, types: Type[]) {
    return new TupleType(types)
  }

  getType () {
    return this.typeArgs[0]
  }

  toString () {
    return `(${this.typeArgs.join(', ')})`
  }
}

export class TypeVariable extends Type {
  static nextId: number
  static nextName: number

  id: number
  instance: Type

  constructor () {
    super('')
    this.id = TypeVariable.nextId ++
  }

  toString () {
    if (this.instance) {
      return this.instance.toString()
    } else {
      if (!this.name) {
        this.name = String.fromCharCode(TypeVariable.nextName ++)
      }

      return this.name
    }
  }
}

// Factory for curried function types that take any number of arguments
export function makeFunctionType (argTypes: Type[], returnType: Type): Type {
  const [firstArgType, ...tailArgTypes] = argTypes

  if (tailArgTypes.length === 0) {
    return new FunctionType(firstArgType, returnType)
  } else {
    return new FunctionType(firstArgType, makeFunctionType(tailArgTypes, returnType))
  }
}

// TODO
// > Contextual names for TypeVariables, so that many tests in the same
//   JS process can start type variables at A
// > Make nice names for type variables once we have created a lot. E.g. after Z,
//   go to AA.
TypeVariable.nextId = 0
TypeVariable.nextName = 65
