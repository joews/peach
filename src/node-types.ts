import { Type } from './types'

export interface AstNode {
  type: string
}

export interface TypeCheckNode {
  exprType: Type,
  node: AstNode
}

export type Ast = AstNode
export type TypedAst = TypeCheckNode
