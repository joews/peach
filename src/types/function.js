const unify = require('../unify')
const { create } = require('../util')
const { PeachError } = require('../errors')

const ANONYMOUS = 'anonymous'

// Make a user-defined function from an AST node
module.exports = {
  makeFunction (functionNode, parentEnv, visit) {
    const { clauses } = functionNode
    const name = getName(functionNode)

    const [minArity, maxArity] = getArity(functionNode)

    return {
      name,
      minArity,
      maxArity,
      isVariadic: false, // no variadic UDFs in peach yet
      call (...args) {
        for (const { pattern, body } of clauses) {
          const { didMatch, bindings } = unify(pattern, args)
          if (didMatch !== false) {
            const env = create(parentEnv, bindings)

            const [returnValue] = visit(body, env)
            return returnValue
          }
        }

        // TODO in the future this will be unrechable; a complete set of patterns
        //  will be a compile-time requirement.
        throw new PeachError(`λ: ${name} has no clause that matches the given arguments (${args})`)
      },

      toString () {
        // TODO include the source for UDFs
        return `λ: ${name}`
      }
    }
  },

  // Make a built-in function from a JavaScript function
  // JavaScript functions have a single arity, so maxArity === minArity.
  makeNativeFunction (name, jsFunction, minArity = jsFunction.length, isVariadic = false) {
    return {
      name,
      minArity,
      maxArity: minArity,
      isVariadic,
      call: jsFunction
    }
  },

  applyFunction (pFunction, args) {
    if (!pFunction.isVariadic && args.length > pFunction.maxArity) {
      throw new PeachError(`Function ${pFunction.name} was called with too many arguments. It expected at most ${pFunction.maxArity} arguments, but it was called with ${args.length}: ${JSON.stringify(args)}`)
    }

    return (args.length >= pFunction.minArity)
      ? call(pFunction, args)
      : curry(pFunction, args)
  }
}

function getName (node) {
  return node.boundName || ANONYMOUS
}

function getArity (node) {
  const patternLengths = node.clauses.map(clause => clause.pattern.length)
  return [Math.min(...patternLengths), Math.max(...patternLengths)]
}

function call (pFunction, args) {
  return pFunction.call.apply(null, args)
}

function curry (pFunction, appliedArgs) {
  return Object.assign({}, pFunction, {
    minArity: pFunction.minArity - appliedArgs.length,
    maxArity: pFunction.maxArity - appliedArgs.length,
    call: pFunction.call.bind(null, ...appliedArgs)
  })
}
