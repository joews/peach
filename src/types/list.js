// TODO List wrapper type
function isList ({ type }) {
  return type === 'List'
}

// TODO generic value-based equality. This fails for nested lists
//  because the `every` check uses ===. It needs to find the value equality
//  function for the operand types.
function isEqual (listA, listB) {
  return listA.length === listB.length &&
    listA.every((v, i) => v === listB[i])
}

module.exports = {
  isList,
  isEqual
}
