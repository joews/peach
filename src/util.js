// _.extend, but immutable by default
function extend (source, ...extensions) {
  return Object.assign({}, source, ...extensions)
}

// shortcut for creating an object with the given prototype and
//  properties with default behaviour
function create (proto, properties) {
  return Object.assign(Object.create(proto), properties)
}

module.exports = {
  extend,
  clone: extend,
  create
}
