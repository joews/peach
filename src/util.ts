// _.extend, but immutable by default
export function extend (source, ...extensions) {
  return Object.assign({}, source, ...extensions)
}

export const clone = extend

// shortcut for creating an object with the given prototype and
//  properties with default behaviour
export function create (proto, properties = null) {
  return Object.assign(Object.create(proto), properties)
}

export function restAndLast<T> (arr: T[]): [T[], T] {
  const _last = last(arr)
  const rest = arr.slice(0, -1)
  return [rest, _last]
}

export function last<T> (arr: T[]): T {
  return arr[arr.length - 1]
}
