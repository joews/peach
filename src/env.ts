import stdlib from './stdlib'
import { clone, extend } from './util'
import { AstNode, TypedNode, ValueNode } from './node-types'

// Return the default environment for a new program
export function getRootEnv (): RuntimeEnv {
  return clone(stdlib)
}

export function getTypeEnv (valueEnv: RuntimeEnv) : TypeEnv {
  return Object.keys(valueEnv).reduce((env, name) => {
    env[name] = extend(valueEnv[name], {
      exprType: valueEnv[name].exprType
    })
    return env
  }, {})
}

export type TypeEnv = { [name: string]: TypedNode<AstNode> }

// TODO
export type RuntimeEnv = { [name: string]: ValueNode }
