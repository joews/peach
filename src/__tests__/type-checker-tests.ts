/* eslint-env jest */
import parse from '../parser'
import typeCheck from '../type-checker'
import { getRootEnv, getTypeEnv } from '../env'
import { fixture } from './helpers'
import { clone } from '../util'
import { Type } from '../types'
import { TypedNode, AstNode } from '../node-types'

import {
  TypeVariable,
  TypeOperator,
  FunctionType,
  NumberType,
  BooleanType,
  TupleType
} from '../types'

function testTypeCheck (source: string, env = getTypeEnv(getRootEnv())) {
  test(source, () => {
    const parsed = parse(source)
    const [lastNode] = typeCheck(parsed, env)
    const type = lastNode.type.toString()
    expect(type).toMatchSnapshot()
  })
};

function testFails (source: string, env = getTypeEnv(getRootEnv())) {
  test(source, () => {
    const parsed = parse(source)
    expect(() => typeCheck(parsed, env)).toThrowErrorMatchingSnapshot()
  })
}

function testFixture (fixtureName: string, env = getTypeEnv(getRootEnv())) {
  test(fixtureName, () => {
    const parsed = parse(fixture(fixtureName))
    const [lastNode] = typeCheck(parsed, env)
    const type = lastNode.type.toString()
    expect(type).toMatchSnapshot()
  })
};

testFixture('fibonacci.peach')
testFixture('function.peach')
testFixture('str.peach')
testFixture('tail-recursion.peach')
testFixture('array-destructure.peach')

// literals
testTypeCheck(`1`)
testTypeCheck(`true`)
testTypeCheck('`the`')

// def
testTypeCheck(`
  x = 1
  y = true
  x
  y
`)

testTypeCheck(`
  x = \`arf\`
  x
`)

// lambda
testTypeCheck(`(a => 1)`)

// array
testTypeCheck(`[1, 2, 3]`)
testFails(`[1, 2, false]`)

// if
testTypeCheck(`if (true) 1 else 2`)

// branches evaluate to different types
testFails(`(if (true) 1 else \`two\`)`)

// non-boolean condition
testFails(`if (1) 1 else 2`)

// test with some pre-defined types and symbols
// run adapted version of Rob Smallshire's python implementation for a sanity test:
// http://smallshire.org.uk/sufficientlysmall/2010/04/11/a-hindley-milner-type-inference-implementation-in-python/
// (adapted because Peach doesn't have `let` yet; use `def` and test in the enclosing environment)

// the peach type checker works over nodes with an `type` property.
// helper for creating nodes for synthetic type tests
function typed (type: Type): TypedNode {
  return {
    type: type,
    kind: 'String',
    value: 'dummy'
  }
}

// simulate a user-defined type
class PairType extends TypeOperator {
  constructor (a: Type, b: Type) {
    super('*', [a, b])
  }

  static of (name: string, [a, b]: Type[]) {
    return new PairType(a, b)
  }

  toString () {
    return `[${this.typeArgs.join(this.name)}]`
  }
}

const A = new TypeVariable()
const B = new TypeVariable()
const AB = new PairType(A, B)

const testEnv = () => ({
  // No unit type yet, so all functions take exactly one argument
  // Number -> Number
  inc: typed(new FunctionType(NumberType, NumberType)),

  // // Number -> Boolean
  zero: typed(new FunctionType(NumberType, BooleanType)),

  add: typed(new FunctionType(NumberType, new FunctionType(NumberType, NumberType))),
  sub: typed(new FunctionType(NumberType, new FunctionType(NumberType, NumberType))),

  // Number -> Number -> Boolean
  addZero: typed(new FunctionType(NumberType, new FunctionType(NumberType, BooleanType))),

  // A -> B -> A*B
  pair: typed(new FunctionType(A, new FunctionType(B, AB)))
})

// testTypeCheck(`inc`, testEnv())
testTypeCheck(`(inc 1)`, testEnv())

testTypeCheck(`zero`, testEnv())
testTypeCheck(`(zero 0)`, testEnv())
testTypeCheck(`(addZero 1)`, testEnv())
testTypeCheck(`((addZero 1) 2)`, testEnv())

testTypeCheck(`(pair 1)`, testEnv())
testTypeCheck(`((pair 1) 2)`, testEnv())
testTypeCheck(`
  id = x => x
  (id 3)
  (id \`hello\`)
`, testEnv())

testTypeCheck(`
 fib =
    0 => 1
    1 => 1
    x => ((add (fib ((sub x) 1))) (fib ((sub x) 2)))
`, testEnv())

// type mismatch in recursive call
testFails(`
  fibc =
    0 => 1
    1 => 1
    x => ((add (fibc true)) (fibc ((sub x) 3)))
`, testEnv())

// type mismatch in patterns
testFails(`0 => 0 true => 1`, testEnv())

// type mismatch in return
testFails(`0 => 0 1 => true`, testEnv())

// type mismatch in arguments to `x`, because function arguments are non-generic
testFails(`x => ((pair (x 3)) (x true))`, testEnv())

// polymorphic function call
testTypeCheck(`
  id = x => x
  ((pair (id 3)) (id true))
`, testEnv())

// recursive function arguments can't be unified with no further type information
testFails(`f => (f f)`)

// ...but it's ok when there is more information available
testTypeCheck(`
  g = f => 5
  (g g)
`)

// // TODO I can't run this test from Rob Smallshire's implementation
// yet because function bodies are a single expression and Peach doesn't have
// `let` yet.
// generic and non-generic variables
// testTypeCheck(`
//   (g =>
//     (def f x => g)
//     ((pair (f 3)) (f true))
//   )
// `, testEnv())

// function composition
testTypeCheck(`f => g => arg => (g (f arg))`)

// destructured aguments
testTypeCheck(`[1|list] => list`)
testTypeCheck(`([1|[2|t]] => t)`)
testTypeCheck(`([x|[y|t]] => [[x, y], t])`)
testFails(`([1|[true|t]] => t)`)
testFails(`([h|[true|t]] => [[h, 1], t])`)

// curried function calls
testTypeCheck(`
  list = (a, b, c) => [a, b, c]
  (list 1)
  (list 1, 2)
  (list 1, 2, 3)
  ((list 1) 2, 3)
  (list true, false, true)
`)

testFails(`
  list = (a, b, c) => [a, b, c]
  (list 1, true)
`)

testTypeCheck(`
  f =
    (1, 2) => [9, 9]
    (a, b) => [a, b]

  (f 1)
  (f 1, 2)
  ((f 1) 2)
  (f 3)
  (f 3, 4)
  ((f 3) 4)
`)

// FIXME #6 - extra parens needed to immediately invoke a no-arg function
testTypeCheck('((() => `look ma no args`))')

// Functions with multi expression bodies
testTypeCheck(`
1 => 2
x => {
  y = 3
  f = a => 2
  (f y)
}`)

// defs can't use already-bound names
testFails(`
x => {
  x = 3
  x
}`)

// tuples
testTypeCheck(`t(1, 2)`)
testTypeCheck(`t()`)
testTypeCheck('t(t(), t(`hi`, 1, x => t(x, (+ x, 1))))')
