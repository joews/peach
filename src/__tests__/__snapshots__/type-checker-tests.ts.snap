// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
  f =
    (1 2) => [9 9]
    (a b) => [a b]

  (f 1)
  (f 1 2)
  ((f 1) 2)
  (f 3)
  (f 3 4)
  ((f 3) 4)
 1`] = `
Array [
  "Number -> Number -> Array<Number>",
  "Number -> Array<Number>",
  "Array<Number>",
  "Array<Number>",
  "Number -> Array<Number>",
  "Array<Number>",
  "Array<Number>",
]
`;

exports[`
  fibc =
    0 => 1
    1 => 1
    x => ((add (fibc true)) (fibc ((sub x) 3)))
 1`] = `"Type error: type mismatch"`;

exports[`
  g = f => 5
  (g g)
 1`] = `
Array [
  "M -> Number",
  "Number",
]
`;

exports[`
  id = x => x
  ((pair (id 3)) (id true))
 1`] = `
Array [
  "K -> K",
  "L",
]
`;

exports[`
  id = x => x
  (id 3)
  (id \`hello\`)
 1`] = `
Array [
  "J -> J",
  "Number",
  "String",
]
`;

exports[`
  list = (a, b, c) => [a b c]
  (list 1 true)
 1`] = `"Type error: type mismatch"`;

exports[`
  list = (a, b, c) => [a b c]
  (list 1)
  (list 1 2)
  (list 1 2 3)
  ((list 1) 2 3)
  (list true false true)
 1`] = `
Array [
  "R -> R -> R -> Array<R>",
  "Number -> Number -> Array<Number>",
  "Number -> Array<Number>",
  "Array<Number>",
  "Array<Number>",
  "Array<Boolean>",
]
`;

exports[`
  x = \`arf\`
  x
 1`] = `
Array [
  "String",
  "String",
]
`;

exports[`
  x = 1
  y = true
  x
  y
 1`] = `
Array [
  "Number",
  "Boolean",
  "Number",
  "Boolean",
]
`;

exports[`
 fib =
    0 => 1
    1 => 1
    x => ((add (fib ((sub x) 1))) (fib ((sub x) 2)))
 1`] = `
Array [
  "Number -> Number",
]
`;

exports[`
1 => 2
x => {
  y = 3
  f = a => 2
  (f y)
} 1`] = `
Array [
  "Number -> Number",
]
`;

exports[`
x => {
  x = 3
  x
} 1`] = `"x has already been defined"`;

exports[`((() => \`look ma no args\`)) 1`] = `
Array [
  "String",
]
`;

exports[`((addZero 1) 2) 1`] = `
Array [
  "G",
]
`;

exports[`((pair 1) 2) 1`] = `
Array [
  "I",
]
`;

exports[`([1|[2|t]] => t) 1`] = `
Array [
  "Array<Number> -> Array<Number>",
]
`;

exports[`([1|[true|t]] => t) 1`] = `"Type error: type mismatch"`;

exports[`([h|[true|t]] => [[h 1] t]) 1`] = `"Type error: type mismatch"`;

exports[`([x|[y|t]] => [[x y] t]) 1`] = `
Array [
  "Array<Q> -> Array<Array<Q>>",
]
`;

exports[`(a => 1) 1`] = `
Array [
  "E -> Number",
]
`;

exports[`(addZero 1) 1`] = `
Array [
  "F",
]
`;

exports[`(if (true) 1 else \`two\`) 1`] = `"Type error: type mismatch"`;

exports[`(inc 1) 1`] = `
Array [
  "Number",
]
`;

exports[`(pair 1) 1`] = `
Array [
  "H",
]
`;

exports[`(zero 0) 1`] = `
Array [
  "Boolean",
]
`;

exports[`[1 2 3] 1`] = `
Array [
  "Array<Number>",
]
`;

exports[`[1 2 false] 1`] = `"Type error: type mismatch"`;

exports[`[1|list] => list 1`] = `
Array [
  "Array<Number> -> Array<Number>",
]
`;

exports[`\`the\` 1`] = `
Array [
  "String",
]
`;

exports[`0 => 0, 1 => true 1`] = `"Type error: type mismatch"`;

exports[`0 => 0, true => 1 1`] = `"Type error: type mismatch"`;

exports[`1 1`] = `
Array [
  "Number",
]
`;

exports[`array-destructure.peach 1`] = `
Array [
  "Array<B> -> B",
  "Array<C> -> C",
  "Array<D> -> D",
  "Array<Number>",
  "Array<Number>",
]
`;

exports[`f => (f f) 1`] = `"Type error: recursive unification"`;

exports[`f => g => arg => (g (f arg)) 1`] = `
Array [
  "(N -> O) -> (O -> P) -> N -> P",
]
`;

exports[`fibonacci.peach 1`] = `
Array [
  "Number -> Number",
  "Array<Number>",
]
`;

exports[`function.peach 1`] = `
Array [
  "A -> A",
]
`;

exports[`if (1) 1 else 2 1`] = `"Type error: type mismatch"`;

exports[`if (true) 1 else 2 1`] = `
Array [
  "Number",
]
`;

exports[`str.peach 1`] = `
Array [
  "Array<String>",
]
`;

exports[`tail-recursion.peach 1`] = `
Array [
  "Number -> Number -> Number",
  "Number -> Number",
  "Number",
]
`;

exports[`true 1`] = `
Array [
  "Boolean",
]
`;

exports[`x => ((pair (x 3)) (x true)) 1`] = `"Type error: type mismatch"`;

exports[`zero 1`] = `
Array [
  "Number -> Boolean",
]
`;
